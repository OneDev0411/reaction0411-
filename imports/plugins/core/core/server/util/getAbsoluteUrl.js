import url from "url";
import ReactionError from "@reactioncommerce/reaction-error";

/**
 * @name getAbsoluteUrl
 * @summary Returns the absolute/base URL for where a request was made, using the request's headers
 * @param {Object} requestHeaders Headers sent with original request
 * @param {String} requestHeaders.host Domain (including port) request was made to
 * @param {String} requestHeaders.origin URL request originated from
 * @returns {String} protocol://hostname[:port]/, with protocol being from the origin URL
 */
export default function getAbsoluteUrl(requestHeaders) {
  const { host = "", origin = "" } = requestHeaders;

  if (host === "") {
    throw new ReactionError("invalid-parameters", "getAbsoluteUrl requires 'host' from request headers");
  }
  if (origin === "") {
    throw new ReactionError("invalid-parameters", "getAbsoluteUrl requires 'origin' from request headers");
  }

  const originParsedUrl = url.parse(origin);
  const { protocol } = originParsedUrl;

  return `${protocol}//${host}/`;
}
