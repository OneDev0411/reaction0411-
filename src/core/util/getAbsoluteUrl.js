/**
 * @name getAbsoluteUrl
 * @summary Combines and returns the given root URL and path
 * @param {String} rootUrl URL ending with /
 * @param {String} path Path that may or may not start with /
 * @returns {String} Full URL
 */
export default function getAbsoluteUrl(rootUrl, path = "") {
  // Check if the path is already absolute
  const regExp = /^https?:\/\/|^\/\//i;
  if (regExp.test(path)) {
    return path;
  }

  let pathNoSlash = path;
  if (path.startsWith("/")) {
    pathNoSlash = path.slice(1);
  }

  return `${rootUrl}${pathNoSlash}`;
}
