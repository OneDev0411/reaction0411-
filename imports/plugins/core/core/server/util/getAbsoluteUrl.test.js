import getAbsoluteUrl from "./getAbsoluteUrl";

test("throws error if no host header is provided", () => {
  const requestHeaders = {
    origin: "http://localhost:3000"
  };

  expect(() => getAbsoluteUrl(requestHeaders)).toThrowErrorMatchingSnapshot();
});

test("throws error if no origin header is provided", () => {
  const requestHeaders = {
    host: "localhost:3000"
  };

  expect(() => getAbsoluteUrl(requestHeaders)).toThrowErrorMatchingSnapshot();
});

test("returns the correct absolute url", () => {
  const requestHeaders = {
    origin: "http://localhost:3000",
    host: "localhost:3000"
  };

  expect(getAbsoluteUrl(requestHeaders)).toBe("http://localhost:3000/");
});
