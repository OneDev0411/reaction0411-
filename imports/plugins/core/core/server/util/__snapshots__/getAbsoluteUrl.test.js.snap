// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throws error if no host header is provided 1`] = `"getAbsoluteUrl requires 'host' from request headers"`;

exports[`throws error if no origin header is provided 1`] = `"getAbsoluteUrl requires 'origin' from request headers"`;
